### ISSUE TITLE
OAuth example

### ISSUE THREAD

**Issue by seanhess** [seanhess/hyperbole#2]
Demonstrate how to integrate with an OAuth provider. Add library code for it?

**Comment by seanhess** [comment]
The OAuth playground looks like it will work for an example: https://www.oauth.com/playground/authorization-code.html

**Comment by adithyaov** [comment]
Hi @seanhess ,

This library looks like an amazing addition to the Haskell ecosystem. In the spirit of the majestic monolith, I've wanted to use something like this for a long time.

I played around with HTMX and found it quite hard to maintain, but looking at the code samples, it seems this library reduces the maintenance burden I experienced with HTMX.

Before using it in my own projects, I'd like to experiment with it, and what better way to do so than by adding a real-world example.

I'd like to take up this issue if no one else is working on it.

I will write a detailed proposal for how the example would look and seek feedback before I begin working on it.

**Comment by adithyaov** [comment]
> But you will get to know the server requests and Hyperbole effect well.

That sounds like a good start!

I'll go through the `examples` package and put down what I have in mind on this issue.
I can start working on it after your feedback.

**Comment by adithyaov** [comment]
Hi @seanhess 

Plan of Action:

Please let me know if you'd like to suggest any changes. The examples are very well
written, so most of the hard work w.r.t design is already done. The implementation
itself should be straightforward.

**Flow**

I would want the details of the OAuth client initially, and so the flow would
start with a page that accepts a `client_id` and `client_secret`.

Attached also will be some text on how to get these details for
testing. Essentially following the instructions on:
www.oauth.com/playground/authorization-code.html

Once the user enters the `client_id` and `client_secret`, the actual OAuth flow
would start.

There will be a button on the page named "Authorize" which the user can click to
start the flow. This will redirect them to the login page.

Once the user logs in with the dummy credentials, they will be redirected back
to the app, where the following details will be visible:
```
{
  "token_type": "Bearer",
  "expires_in": 86400,
  "access_token": "...",
  "scope": "photo offline_access",
  "refresh_token": "..."
}
```

**Implementation**

I want to add this as a dependency for convenience:
https://hackage.haskell.org/package/hoauth2
I don't think I'll need any more dependencies for this.

This entire flow will have 1 redirection for the auth token and 1 request for
the access token.

I will create an `Example.Page.OAuth` for the code related to this to live in.
I'll need to add another constructor to `AppRoute` called `OAuth`.

**Comment by adithyaov** [comment]
> We could consider pre-registering the app and including the client_id and client_secret in ENV, so that the example looks more like a production OAuth implementation.

 Yes, we should do this. I'll need help in making sure a few `ENV` vars are
populated:

1. `OAUTH_CLIENT_ID`
2. `OAUTH_CLIENT_SECRET`
3. `OAUTH_USERNAME`
4. `OAUTH_PASSWORD`

This is deployment-specific, I believe.

I'll read these `ENVs` from the environment. Please let me know if you have any
suggestions on the `ENV` variable names.

`OAUTH_USERNAME` & `OAUTH_PASSWORD` are required to show dummy credentials to
the user running the examples. Since they are tightly coupled with the
`OAUTH_CLIENT` keeping all this information together seems like a good idea.

> The main page should show whether they are logged in or not, and survive a browser refresh. To do this, you'll need to store the auth token in a session.

> They should be able to log out

Yes, this makes the example much more elegant!

I think I have all the information to get started. Will open a PR soon.

Best,
Adithya

**Comment by seanhess** [comment]
Done! See #116
